### ISSUE TITLE
Concept exploitable?

### ISSUE THREAD

**Issue by Pleezon** [Minecraft-Java-Edition-Speedrunning/mcsr-serversiderng-1.16.1#9]
Hello, I might have found a serious issue with the mod.
Full disclaimer: This is my opinion, I might have a faulty thought process.

In its current state, the server supplies a seed from which the client derives seeds for the random generators related to each implemented "random" in-game event. 
Having a separate random generator for each random event does indeed make it easy for a mod to verify a run's integrity by simply checking if the runner's random generator behaves the same way as theirs. 
In my opinion, this approach is fatally flawed.
By fetching the seed the server supplies at the beginning of a run, a runner can simply "calculate their future luck" by generating the stream of events a LRNG with the given seed should produce (similar to how people verifying runs would) prior to running, thus giving them crucial information to decide on whether to request a new seed from the server ("resetting").

tl;dr: With the correct skills, someone could make it so that their drop rates are significantly higher than they should by automatically skipping runs in which the seed isn't in their favor.

One might suggest leaving the game logic as-is; using one LRNG for all random events; But this approach also bears fatal problems for the mod which I will not explain here. 

edit: removed a part since it might be misinterpretable

**Comment by jan-leila** [comment]
the rng seed gets refreshed every 30 seconds

**Comment by Pleezon** [comment]
The server's endpoint for refreshing the seed (basically the login endpoint, I believe) currently has no rate limit. 

(at least it seems like it)

**Comment by Pleezon** [comment]
Also, one would also be able to cheat without performing additional requests.
Just use the seed to calculate how much gold is needed for the next piglin trades to result in a specific amount of pearls, and é voilla, time saved by not having to gather as much gold as one normally would.
This is just an example, I have no clue how speedrunning works.
I'm just trying to illustrate how this mod could be abused.

**Comment by jan-leila** [comment]
and you need to be able to pull that all off in the next 30 seconds or else you have all new rng, and with that its still a better scenario then letting the player pick their rng

you could wait around for good rng if you wanted to but the point of speedrunning is you didnt wait around

**Comment by Pleezon** [comment]
It might be worthwhile investigating what would happen if the mod were to only use one random generator for all relevant random events.
That way, the outcome of the game, e.g. the amount of piglin trades, is entirely dependent of player behavior.

**Comment by jan-leila** [comment]
that would be worse, then you could trigger random events to skip bad calls letting the player directly manipulate the rng instead of predicting it for a short amount of time

**Comment by Pleezon** [comment]
> and you need to be able to pull that all off in the next 30 seconds
> you could wait around for good rng if you wanted to

What I'm saying is that a program could cycle through thousands of them in a couple of seconds or minutes

**Comment by jan-leila** [comment]
thats not an issue on this repo go to the backend repo for that issue

**Comment by Pleezon** [comment]
where's that?

**Comment by Pleezon** [comment]
I'd be interested in looking at the backend code, but I cant seem to find the repo. Is it public?

**Comment by jan-leila** [comment]
sorry had a work meeting the repo is here https://github.com/VoidXWalker/serverside-rng-website, nvm looks like its private we know about the issue now though and will try to get it fixed asap though

**Comment by jan-leila** [comment]
pls dont dos the server again btw, everything went down for like an hour this morning because of you making to many requests, and we do log the time that the requests are made so we would know if someone exploited this anyways by looking at the request volume
